distribution: Deployment pipeline UX goal (#1787)
sourcegraph/about: handbook/engineering/distribution/goals.md

## Medium-term goals (~3-6 Months)

### [Improve deployment pipeline UX](https://github.com/orgs/sourcegraph/projects/96)

- **Owner**: TBD
- **Status**: Not started. Unknown amount of work.
- **Outcomes**:

>   - Its simple to identify if a commit has been deployed to an environment

Reviewer: ```suggestion
  - It is simple to identify when a commit has been deployed to an environment
```

I feel like _if_ is a bit too easy to solve (for example, you can already kind of tell from the build version in `https://sourcegraph.com/users/$ME/settings` ), but much more interesting would be knowing _when_ a commit has been deployed (for example, via a comment on the PR, or through Buildkite, or Slack, or etc.) which seems like it could be nice for triage as well (a question that seems to come up often is, "did something deploy at x time?")

Reviewer 2: Yeah, agree, plus, I'd like to see the history of deployments easily.

Author: >I feel like if is a bit too easy to solve 

Thanks, yeah that is what I meant, not only when as in "timeline" but rather, I merged, is my commit already in the environment?

>like to see the history of deployments easily.

Lets add that!

Author: Unresolving to highlight the conversation

>   - We can trigger a deployment via a `/` command in Slack

Reviewer: I was under the impression that the `/`-command thing was primarily for fast rollbacks, which I think is a nice idea - I think `/`-command just for deployment seems like a crutch for what should be a more transparent deployment process (what we should aim for, ie the point above, which should probably encapsulate "how my commit gets deployed is obvious")

Author: It depends on how you look at it, I think a simpler https://github.blog/2015-06-02-deploying-branches-to-github-com/ is highly desirable and gives a lot of visibility onto what is deploying, when it finishes, etc.

Reviewer: I think we can the visibility on deployment while keeping the automated fashion of "what's in master is what is out there" (which I think is a lot less confusing than the model in the blog post, which is "any branch could be out there" - I'm also curious about whether they still use that model, given the post is 5 years old and some of the repos linked are now archived 🤔 )

Author: I dont know if they do, and I would not map it 1 to 1, I think a simplified version of that works quite well. Its similar to what was shared here https://technology.riotgames.com/news/how-riot-games-uses-slack

Reviewer 3: 👍 from me on keeping it automatic. 

I think we should strive to have anything committed to `main` be ready for production and therefore can be deployed to any environment. We can use feature flags to actually enable / disable it per environment. For example if we want to demo a new feature on demo.sourcegraph.com before it's public.

Reviewer 3: I'm a bit torn between whether we actually need a `/`command. If deployment is automatic then I don't think we need it for that. Fast rollbacks are nice, but it means that we are out of sync with what is committed vs what is deployed. If CI was fast enough (I know that's a stretch) then a rollback could be done via a git revert.

Author: There are 2 types of rollback from my point of view.

**Roll forward**: Revert the code, but you are actually just creating a new artifact that removes a feature/change
**Roll back**: Redeploy an existing artifact

Reviewer 3: In that case in an ideal world I would prefer Roll Forward 😄

Reviewer 4: Roll back seems like the same thing, but is "instant" vs needing any artifacts to be built. I don't think roll forward is achievable.

Note: our infrastructure is in a repo. So we will always be tracking the `release` branch of that. So for example a roll back would a roll forward in the infra repo :P

Reviewer 2: A roll back could also just be a direct git revert without a PR?

Author: IMO its still a "roll forward" as you have to rebuild the artifacts, you are creating a new version that lacks a feature, not restoring an older version.

## Medium-term goals (~3-6 Months)

> ### [Improve deployment pipeline UX](https://github.com/orgs/sourcegraph/projects/96)

Reviewer 5: Is there a problem statement beyond the negation of the desired outcomes?

Author: There is, I forgot to add it. Ill create another PR to include that.
